### PR TITLE
[Testers Needed] vm_native.cpp: Remove the blame from Windows

### PR DESCRIPTION
**Pull request by elad335**
Use the Windows 10 memory mapping API which has the functionality required for preserving memory space for mappings.
Previously RPCS3 hacked it by releasing the reserved memory space temporarily and then mapping file mapping. But this induces a horrible race in which another memory reserving call in another thread can take the placement of the file mapping when the memory space is completely free for everyone to take possession of.
Fixes this error: "**Memory mapping failed - blame Windows**" on Windows 10+.

### CODE REVIEW

**Comment by Satan86** [comment]
GT5 crashes on boot with this PR
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/8839707/RPCS3.zip)

**Comment by elad335** [comment]
Thanks, retest.

**Comment by Nekotekina** [comment]
What's with all these changes in DYNAMIC_IMPORT, looks pointless.

**Comment by elad335** [comment]
It's because it's called in vm::g_base_addr initialization in vm.cpp and vm_native.cpp static variables have yet to be initialized.

**Comment by Satan86** [comment]
Game didn't crash on boot, ran fine until Desync

**Comment by Nekotekina** [comment]
@elad335 dynamic import variables are constant initialized. Could be marked as constinit.

**Comment by elad335** [comment]
Fixed.

File: Utilities/lockless.h [hunk 1]
```diff
@@ -47,6 +47,19 @@ class lf_array
 		// Access recursively
 		return (*m_next)[index - N];
 	}
+
+	u64 size() const
+	{
+		u64 size_n = 0;
+
+		for (auto ptr = this; ptr;)
```
Comment (Megamouse) [inline Utilities/lockless.h:55]: Nitpick
```suggestion
		for (auto ptr = this; ptr; ptr = ptr->m_next)
```

File: rpcs3/Emu/Memory/vm.cpp [hunk 2]
```diff
@@ -23,11 +23,11 @@ void ppu_remove_hle_instructions(u32 addr, u32 size);
 
 namespace vm
 {
-	static u8* memory_reserve_4GiB(void* _addr, u64 size = 0x100000000)
+	static u8* memory_reserve_4GiB(void* _addr, u64 size = 0x100000000, bool is_memory_mapping = false)
```
Comment (Megamouse) [inline rpcs3/Emu/Memory/vm.cpp:26]: this looks like it could be a template instead

File: rpcs3/util/vm_native.cpp [hunk 3]
```diff
@@ -87,6 +89,97 @@ namespace utils
 #ifdef _WIN32
 	DYNAMIC_IMPORT("KernelBase.dll", VirtualAlloc2, PVOID(HANDLE Process, PVOID Base, SIZE_T Size, ULONG AllocType, ULONG Prot, MEM_EXTENDED_PARAMETER*, ULONG));
 	DYNAMIC_IMPORT("KernelBase.dll", MapViewOfFile3, PVOID(HANDLE Handle, HANDLE Process, PVOID Base, ULONG64 Off, SIZE_T ViewSize, ULONG AllocType, ULONG Prot, MEM_EXTENDED_PARAMETER*, ULONG));
+	DYNAMIC_IMPORT("KernelBase.dll", UnmapViewOfFile2, BOOL(HANDLE Process, PVOID BaseAddress, ULONG UnmapFlags));
+
+	const bool has_win10_memory_mapping_api() { return VirtualAlloc2 && MapViewOfFile3 && UnmapViewOfFile2; }
```
Comment (Megamouse) [inline rpcs3/util/vm_native.cpp:94]: why no newlines ?

File: Utilities/lockless.h (hunk 1 above)
Comment (elad335) [reply Utilities/lockless.h]: Fixed.

File: rpcs3/util/vm_native.cpp (hunk 3 above)
Comment (elad335) [reply rpcs3/util/vm_native.cpp]: Fixed.

File: Utilities/lockless.h [hunk 4]
```diff
@@ -47,6 +47,18 @@ class lf_array
 		// Access recursively
 		return (*m_next)[index - N];
 	}
+
+	u64 size() const
```
Comment (Megamouse) [inline Utilities/lockless.h:51]: kinda contradicts the description above now

File: rpcs3/util/vm_native.cpp [hunk 5]
```diff
@@ -87,6 +89,100 @@ namespace utils
...
+			auto& info = s_is_mapping[i];
+
+			if (!info.addr && info.state.compare_and_swap_test(0, 2))
+			{
+				info.addr = addr;
+				info.size = size;
+				info.state = 1;
+				return true;
+			}
+		}
```
Comment (Megamouse) [inline rpcs3/util/vm_native.cpp:179]: Theoretically this misses a definite exit condition.
Is it guaranteed to always find a free spot?

File: rpcs3/util/vm_native.cpp [hunk 6]
```diff
@@ -671,32 +779,72 @@ namespace utils
...
+			}
+
+			if (target + m_size < base + size && !::VirtualFree(target, m_size, MEM_RELEASE | MEM_PRESERVE_PLACEHOLDER))
+			{
+				return {nullptr, "Failed to split allocation end"};
+			}
+
+			if (cow)
+			{
+				// TODO
```
Comment (Megamouse) [inline rpcs3/util/vm_native.cpp:810]: do what?

File: rpcs3/util/vm_native.cpp (hunk 5 above)
Comment (elad335) [reply rpcs3/util/vm_native.cpp:179]: Yes, when the array buffer is all checked, it extends the array with data that must be available for this function.

**Pull request merged**